#!/usr/bin/env bash
#===============================================================================
#
#          FILE: salmon.sh
#
#         USAGE: source salmon.sh
#
#   DESCRIPTION: Analysis with Salmon
#
#        AUTHOR: Rohit Suratekar
#  ORGANIZATION: IIMCB
#       CREATED: Tuesday 20 August 2019 13:30
#      REVISION:  1
#===============================================================================

set -o nounset # Treat unset variables as an error

source config.sh

generate_decoy() {
  # Generates Decoy files which can be used for Salmon Indexing
  #
  # -j : Number of threads
  # -a : Genome Annotation file
  # -g : Genome FASTA file
  # -t : Transcriptom or cDNA dump file
  # -o : Output path

  log "Generation of Decoy files for Salmon indexing started"

  bash ./external/generateDecoyTranscriptome.sh \
    -j 8 \
    -a "${FOLDER_GENOME}/${GENOME_ANNOTATION}" \
    -g "${FOLDER_GENOME}/${GENOME_FASTA}" \
    -t "${FOLDER_GENOME}/${GENOME_CDNA}" \
    -o "${FOLDER_SALMON_INDEX}/"

}

index_salmon() {

  # Generate index for salmon
  #
  # -t : gentrome.fa generated by decoy script
  # -i : index folder
  # -d : decoys.txt generated by decoy script
  # -k : k-mer length. 31 is default and recommended by the authors
  # -p : number of threads

  log "Salmon indexing started"

  "${TOOL_SALMON}/salmon" index \
    -t "${FOLDER_SALMON_INDEX}/${SALMON_HYBRID}" \
    -i "$FOLDER_SALMON_INDEX" \
    -d "${FOLDER_SALMON_INDEX}/${SALMON_DECOYS}" \
    -k 31 \
    -p 8
}

READ_1="${FOLDER_DATA}/${SRA_ID}_1.fastq"
READ_2="${FOLDER_DATA}/${SRA_ID}_2.fastq"
SALMON_OUT="${FOLDER_ANALYSIS}/${SRA_ID}/salmon"

quantify_reads() {

  # Performs quantification with Salmon tool
  #
  # -i : Salmon index folder
  # -l : Library type (A=automatic)
  # -1 : forward read
  # -2 : reversed read
  # -o : Output folder
  # --validateMapping : enables selective alignment of the sequencing reads

  log "Salmon quantification started"

  "${TOOL_SALMON}/salmon" quant \
    -i "$FOLDER_SALMON_INDEX" \
    -l A \
    -1 "$READ_1" \
    -2 "$READ_2" \
    --validateMappings \
    -o "$SALMON_OUT"
}

# Makr folder to keep the index files and output analysis
mkdir -p "$FOLDER_SALMON_INDEX"
mkdir -p "$SALMON_OUT"
# Check if decoy files are available
if (check_file -p "$FOLDER_SALMON_INDEX" -s "+0M" "$SALMON_DECOYS") &&
  (check_file -p "$FOLDER_SALMON_INDEX" -s "+0M" "$SALMON_HYBRID"); then
  log "Decoy files are already available, skipping generating them."
else
  generate_decoy
fi

# Index salmon (can change this in config.sh file)
if [[ ${INDEX_SALMON} -eq 1 ]]; then
  index_salmon
fi

# Start the quantification
quantify_reads
